hibernate-core/src/main/java/org/hibernate/annotations/GenerationTime.java

```diff
@@ -6,46 +6,80 @@
  */
 package org.hibernate.annotations;
 
+import org.hibernate.AssertionFailure;
+import org.hibernate.Internal;
 import org.hibernate.tuple.GenerationTiming;
 
 /**
- * Represents the timing of value generation that occurs in the database.
- * Intended for use with the {@link Generated} annotation.
+ * Represents a class of events involving interaction with the database
+ * that causes generation of a new value. Intended for use with the
+ * {@link Generated} and {@link CurrentTimestamp} annotations.
  *
  * @author Emmanuel Bernard
  *
  * @see Generated
- *
- * @deprecated use {@link GenerationTiming}
+ * @see CurrentTimestamp
  */
-@Deprecated(since = "6.2")
 public enum GenerationTime {
 	/**
-	 * Indicates the value is never generated.
+	 * Indicates that a value is never generated.
 	 */
-	NEVER( GenerationTiming.NEVER ),
+	NEVER,
 	/**
-	 * Indicates the value is generated on insert.
+	 * Indicates that a new value is generated on insert.
 	 */
-	INSERT( GenerationTiming.INSERT ),
+	INSERT,
 	/**
-	 * Indicates the value is generated on update.
+	 * Indicates that a new value is generated on update.
 	 *
 	 * @since 6.2
 	 */
-	UPDATE( GenerationTiming.UPDATE ),
+	UPDATE,
 	/**
-	 * Indicates the value is generated on insert and on update.
+	 * Indicates that a new value is generated on insert and on update.
+	 *
+	 * @since 6.2
 	 */
-	ALWAYS( GenerationTiming.ALWAYS );
+	INSERT_OR_UPDATE,

```

Comment:
I prefer ALWAYS (or BOTH) over a needlessly longer name with underscores.  "Reads" better imo.

It should be AND at the very least, not OR

Reply:
So I struggled quite a lot with this one. The issue is that the name I originally wanted to use, `timing`, was taken, and I had to look for something else. I tried a few different things like `on`, which I quite liked, but none of them read well with `ALWAYS`. And then the name I *really* liked, `event`, reads *really* badly with `ALWAYS`.

And then it sort of hit me that `ALWAYS` is a bit too strong anyway. It's not *always* regenerated, only when the entity is inserted or updated.

So:

-  "generated when `event=INSERT_OR_UPDATE`" or 
- "`@CurrentTimestamp` of the last `INSERT_OR_UPDATE` `event`"

sounded reasonable to me. (But I'm very open to other ideas.)

Reply:
> It should be AND at the very least, not OR

Depends how you read it. If we think its describing a union of sets, then yes, "and" is what most people use to describe a union in everyday speech.

But if we think of it as a predicate applied to a single event (a "filter") then it's "or". (Mathematically, set union actually corresponds to logical disjunction.)

So it it were plural, I would go with "and", as in `INSERTS_AND_UPDATES`. But since it's singular, `INSERT_OR_UPDATE` sounds more correct to me. `INSERT_AND_UPDATE` sounds weird because no event is *both* an insert *and* an update at the same time.